### ISSUE TITLE
How to use `Value.Check` with raw json schemas?

### ISSUE THREAD

**Issue by dswbx** [sinclairzx81/typebox#807]
According to https://github.com/sinclairzx81/typebox/issues/682 it is currently out of scope for typebox to work with raw json schemas, but this request mainly focuses on value-check typebox-generated json schemas (not to get types from it). What would be the best way to do this? I can't use AJV, as cloudflare environment doesn't allow arbitrary code execution. 

I want to validate raw json schemas (generated by typebox) using typebox, because I store the output in database and then retrieve it to validate objects. This won't work:

```ts
import { Type } from "@sinclair/typebox";
import { Value } from "@sinclair/typebox/value";

console.log(
   Value.Check(
      Type.Unsafe({
         type: "object",
         properties: {
            foo: {
               type: "string",
            },
         },
         required: ["foo"],
      }),
      { foo: "bar" },
   ),
); // throws
```

Consulting the code for `Check` I found out that it checks the `Kind` symbol, so modifying it slightly like the following worked:

```ts
import { Type, Kind } from "@sinclair/typebox";
import { Value } from "@sinclair/typebox/value";

console.log(
   Value.Check(
      Type.Unsafe({
         [Kind]: "Object", // added
         type: "object",
         properties: {
            foo: {
               [Kind]: "String", // added
               type: "string",
            },
         },
         required: ["foo"],
      }),
      { foo: "bar" },
   ),
); // works
```

I am tempted to believe that I could "simply" write a recursive function that adds the `[Kind]` prop based on `type` to make this work. Would this work? Or is there a better way to do it? In any case I believe it would be a nice addition since the schema could also be received over the wire (e.g. OpenAPI spec).

**Comment by sinclairzx81** [comment]
@dswbx Hi,

Unfortunately, the `[Kind]` symbol is a requirement to use the TypeBox Check and Compiler functions (as this property is used both for internal runtime optimizations and type composition checks). In saying this, you can still use raw schematics with TypeBox, but you will need to map them into TypeBox types first (as you mentioned)

---

> I am tempted to believe that I could "simply" write a recursive function that adds the [Kind] prop based on type to make this work. Would this work? Or is there a better way to do it? 

So, there's actually been some work done recently to provide better support for raw schematics with a prototype `FromSchema` function prepared at the link below.

https://github.com/sinclairzx81/typebox/blob/master/example/prototypes/from-schema.ts

### FromSchema

The FromSchema function will attempt to map raw schematics into TypeBox types if an appropriate transformation exists. This function has a computed return type of `TSchema` that can be used for type inference and additional type composition. Note that any unrecognized raw schematics will result in `TUnknown`.

[Reference TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo43RuAZRgEMZgBjOAXzgDMoIQ4ByAAQGdgA7CgGxLAoAejyYARhAAeLAFCyKEbh3jI4AXjgAxRiEIUAFhhAkAFEjiWr168OFxFy1QC4UAeXEArDBRgAeBFlLMQxXFggvHxgWHBs4+Lg7KzoICFdkYigeAHMsILgwRkxYYAwOVwsEqqsky3ESKHTMnLybFLTEXHwwlSzubJYaWLjauHqALyaYPuz8q3rGzpCe6ZzB6mGbJOoAPjm6kkml7tZetaH8jfyoDABHAFchDAATVwBtFnaY1gXvlgmWABdPK0EgcBxKFQASnkjhUcAAChoUPgAHQIhrkEh8UzIKHVSxJOHwBHpDzeXwBfYEmnVWrtMlgchKbF+DKrfo7Vq0nlVWoLRnM7is9kzLnU3mSxL2MaHQXAFl8NnNTncqXq7Z7WQhFDI4hkSh+EIQOgoPY6pGafXkChG-AmxFapIAWldbvdHs9Xu9Pt9fu9siSOiY+iMJlcAGEIGBUHASNxnoiwTACABJcB8YwYbikIVwABCGD4EAA7oHhP7K1XqzWPfJQJBYHAAFRx8HIfP0XSsTg8fiCEQhSQyeQu2vjid+8tEQzGA0UDjTyfLlduhSQ+CpjgAUSkJF8yNMADdsfdQnB7twANbcUvcHAYKSYXyuS83u-49Q7OAnvhnjTqJoj7PjA65OHAW6EAAaqeBCaMesGvtet4ltwUKuL+-7AOC0GwRo34dqiMF-hgADi9wNM8qKQRy2QISR+IAD6MSg+ZEbB5GUdRHAAHL3CA4gYFA9FnkxLGEcRZ6cVAVFbvmqSZvGIkYDCxIQRwhDbtw-GHph55vihaEYXh2FEFpOlfqx7EkdJskcOSUTKfiABkzlWZJZEUTJ3EAIJQFAJCoMpqLZvxLluXpIXaSAIVHkJQVOfh6m4SRTmqRuyU+XwfBuKa8F6Uh76oehP4mThWU5XlBFsR5tncQ5viJa57kcV5dl+QFQWRdilXpeBkE+dwqC5bpiEXshH7GSRcCmYQg3DVVLU2W19WRI1enhUtUkrVuHWBcF8YLX18KQW43AYCN+VjQZk2ldNs1nRdi0Sa1XFbg1MBNW5L3LW9HB7V1sGokoT3HZuGnIPcYCZqNJGFYZJV6TNOGQ9DcHVdZ21-R9X3qbu+6fZFIQ4CwDSdSwm0-Vj3m7f5+2RcAKYgBwYOZXTsZXXD41FUZd1YeV7NJVTnnY2thOwZtW74+tQPE6wZOBRTcDNQAhMLdW051wWM8YLPK99NWvTT9li9rTMs2BJ0aVG4Gc2e8O3Ujs02-ClnqztJsUuLDH61tIvGzjekfMSQKs5BKrZLD9vcwjU380QEdC4bv0B6bG2+1Le4yyRqJyywZz9BTlvg4QfECUJUf6RNxVxwQs1l4JUBJ5j-t2YHEsZzuWfe2euf4CT0WN0XamQamOYYNkFd21XPOI2VRBjymk9N27yfU23acd81mcE8Fec8EvQnDxlkHyRAincJXDs13zdc4WfF-N7VHvtz729d7vRP9z8CkYPGx-9Q0nxbKV8Y6O3noQB+f9L6rxbhrT2jl07v2lj3DAfdMAD3uNlABVtCAIiKEJcgZRQE3Rvk7HC+DoyENKOCWBz9RZeyctOQoGAYBEKgM6YA2Rbw3GLslD6JDq683IUQARdCjYb0YUgtyO9s69zzhEL2St354IISUMowVChUPUXrZqwUbgPCeImQCmhLzPAwHQHgLw4DMT9vAgG+i7iPBuM8TakUDHOJeLFeKphlLXzQk-CR3EI5pRhDqFK-5NAF0jixQeFcWKSHPtA7U+AzLRWRAAJT-s8JQfBUABDgKFEArgbgkBydwPJRAPJvEBDQc0qS5rZUunALJZTcn5KQD1XKJTsntLAahGpdSUmYCIPNZprTyl5IKYdbpcBSmTNjKQ7ggzdjDIIIQR64zekVI6XAEGsz5l9KWSs+pIzVHaKIbQlpPhoDPD8JQ4oeAADSGBUA4CWac9ZYjrltJ2QU5YrBFFRBYAAbgKGoy5rhzmPJoWCjxRiAD8PTfmVOiSctZRBUYw00BM9p-yTik3ZqCma5tkULP6cs2pqzwkA0ydsqZSAAWEvJmCnWzN-FDPCS7eAOL6W7OJFCjynKGmJ15Si3ZTLonrE+UQBuU8fkLPxZgMIcSoDSoxYQReE95W4r+YyglB9tVquFWcqB8Y6XiqVeef4v9-4mvWcAvgFrFX6uVawbS2ChljlXD62s05gwgCySoJcvrQ1VmYTcNhpQOFcJ4RgDFyAA1Bv8GoR8KYEzgkOTsilNScA+QoFQNN2ZnjtjYhkWcJhBmaBqd+eC+RU1SHTSWuZfK4BvB4HQCuAAZQpjbi3giWTgVEw6O0Vwyb2ptA6hE1MBPsBFKAA2hjnH4Lt34i0ZpmtwTtTdCATv7VZctYYSASnnYm3Qya-AZJwG8YdqJ80UBwIQQEew4jpCTWUfwV6223vvc+-Yrh72yBhHQS8vgFSX3fSoNke6N1Zsqcc59uJ0glTPUwC9yBvyBEsJG+4UBL7IFRCYMAviu2fm-IuitZhSP4jBHAc6cUoCyGoCGsNrGfT+t0OZQNH6WNsb4+6CNrD2GcO4dAeNOpUMgC4+hmDza4OxgiRgXNcB72ydLaiQ9c4q1tufYeetamW3irbaOpuPb13NsU0OkdW6x0GcUzOudC7OPRQvV+m9w7704EIsgLtOsApKlXX+6wAGC1AdkCB3geYA3SY-dB8zmbW32cQ8gZDb7nP8XQ5h64rDcP4cIyQYjphSPN18ymfzRWoQ0fBPRoSTHeP8f49OCquV6sNbY4JqNQkRNxoTQG5rdA4t9o3Y0yqta4CmHrZBlNHwul0CBGuobzaTPXPhPFg9S7K2zusKestWqoAcCiNuTCZAXiXo-S+qw6Qy08QwAxsLEWwNKG0Lofrg3J2jKaXQHYSGUAob659tkWXsM5bw1ZPbB3fBHdPCd54pgpu4lRLNqEXmYTMe9W1vjTWhotfRxj1jHXhOxrE71l72OBsNve3NMnY2JuWEk+hmbZP5sGeW8mgz3mNskBqY57zABVbg4God-hh2dlQF3LBXY0zdu7wHQNRdJwtN7+6qcLW+yl37aWmBjPJ0DltMBctWX54L47KZYfw4IzMugyPft1dx3j0N05Nl0Fa-bn1BPo3deJxJgNTulfDadzTyb57YvIA+Psubum1us4-ezstnPufbYPUbpQQuKKm9FzAcXcBJfIGl7V2XkXwPPaYL7inyuA8-bxJrkApfdc4dB4RZP3BU8w7h8HlQCPw-W7xLb+wrv2tJC4y7-vy53ddaJ7w736WQB+4s1xwPdPosuZDx8IpzOo82abtuTMIBswwHBGtjnlGE9WB2xppvLf0-b6zDmDgWec958YwXx7EHp+z5wvPyvf238YcQNl-XDebEF+JuLwbeTAMWHeBGRSlWvecAI+Ya04WK8adu8Bk4Y+Maomk+qSkmSB7+mKUMmYC+TmaGK+LAbKHA6+i24Iy21+u+t+seGm8eW2p+B6uBtBe+d+-6rBBBGAfgNaYWSQLCnWGBPWD28uTAuBZew2SBauqWxBIAkhdeIO+GZaPBYB3GkBqI5B3eVWdGt2tWaOfeqBbuSQAMw+xhNY6BnuWBIykmAMeBc07MRBkmnObIHw5BlB72y2qY5sDBmmJgPOZa9hPhus9+B69hfOQiewMIghkahOmB8aYhRefW7MDhAMshGu8hERShABKhd67M6hnOCO2hKObYehd2hhcBFhK4043K5h1RlYVhE+4m2BAa3KDh3KzhbEWkHAuGvB3mpWQkIowcG4820R4WcuyRug7RUhFmnRX+f+wOuRVkgx5WBGxIKOsBDRo+SQH09R2xvoJOTADy1CZQ70TI4GrITyBmJxsAqALybyLSBm0SNiFKY2SyBm46860w-4rg1xa2Xx9A2IB22euAUAZ4TRCRRxIAtxlyDhsJNCOA46a2LxLEHyyIhEl+vBWGcB8mba1xPAcAV4ryDoyAgI1eCJZx9kFxiofgTySJC272Px8aie3mbgNJwoSoLhlGbhTyz6YR3JR6vJf6qysRQmHuzR0JH0DhH0zhAalJHAbhLAWiMKZQQIOAu4-A9w5iSp8KLi6p405ili50zwOwbwqqkeVBm626iIEKNCfhCpgRGm0pCpApZa0pCAqyz+4hNeYsMpYsGRVe8h0pv+OJakKppxVy2QxY9QfAyABg2EqIH0qIk8MAbgqEsJqAPEJAu+HACOEZOiUIqILi9wFAGAvinmcATyZGixVg9el8SAP6Ba16fJ6QxZTiRivsBGepXifYf45ieZ1ZcA86FGR6+ZdpZQbwfJ+Iku7JQo2IhRlG45FyNCU5gIlWNAlwOAnpZR3mCpMISxBuhEOMBZlypRtGNWjGlRBxE4HGIYlG+xN5XokJPWwE0A8AU+95Qpv+dadOdm-Ww5WRAOv+zg+ma2KuI0p6fWZOgO2eYFVpGy50kFwZSFOucFf54FXGgFkmXGsFoFGFCFAiUFugIZ34+Fuo4FSB2FAaih6FFFCFtKxFWuqRIF8FlO3K1F0xG4eFbFyuicZ+ooOQdFsxOEcqTcAlYlwldmWqy82Fu248sl5FIlRAZql8AlqlUl4FjqclUuWCTqSlZa-OPMAhzYrYGI+2xCJAKImA8kUgXQIyDATAcyJAJYcAAAUhwEoJzi2MILIOYv2DcPQJMU9qOcuhhl-tXq4b-qOPYPcCCaFSYH4GpL-k9hkHAAAKyogAAM04-lAggVSRIVugrhyVgZ3+X5YV34QAA)

```typescript
import { FromSchema } from './prototypes/from-schema' // copy to your project
import { Type, type Static } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

const T = FromSchema({               // const T: TObject<{
  type: 'object',                    //   foo: TString,
  properties: {                      //   bar: TString,
     foo: { type: 'string' },        //   baz: TString
     bar: { type: 'string' },        // }>
     baz: { type: 'string' },
  },
  required: ['foo', 'bar', 'baz'],
} as const)

const P = Type.Partial(T)            // const P: TObject<{
                                     //  foo: TOptional<TString>,
                                     //  bar: TOptional<TString>,
                                     //  baz: TOptional<TString>,
                                     // }>

console.log(Value.Check(T, { foo: 'hello', bar: 'world' })) // true
```
So FromSchema is currently the recommended implementation provided by TypeBox, but isn't ready for inclusion in the library yet (but will likely be included when TB drops support for TS 4.0). For now, you can copy and paste the `from-schema.ts` module into your project and modify the code based on your requirements.

Hope this helps!
S

**Comment by sinclairzx81** [comment]
@dswbx Hey, good to hear :)

> Out of curiosity, do you have a timeframe in mind for when you'll drop support for TS 4.0?

Nothing formal as of yet. TS 4.9.5 is still pulling around 20% of all TS downloads (currently 9,765,927 weekly) so will want to see that download figure go down a bit more before dropping support in TB. 

Note when 4.x support is dropped, I don't anticipate any major changes in TypeBox when that happens. I'm mostly looking to transition over to [Const Type Parameters](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#const-type-parameters) where appropriate which should be a non-breaking change for anyone on TS 5.x. Unfortunately I can use it today without potentially breaking 4.x users :(

With this said, you can use this feature on FromSchema if you update the signature to.

```typescript
export function FromSchema<const T>(T: T): TFromSchema<T>
//                         ^ add this
```
As a rough estimate on when 4.x will be dropped tho, probably sometime nearer the end of this year (as a guess) and after I've had some time to upgrade some of my build tooling (which is currently on older versions of esbuild that also doesn't support const generics).

Hope this brings some insight! Will close of this issue for now.
Cheers
S
